src/arfs/multi_chunk_tx_uploader.ts

```diff
@@ -0,0 +1,253 @@
+interface MultiChunkTxUploaderConstructorParams {
+	gatewayUrl: URL;
+	transaction: Transaction;
+	maxConcurrentChunks?: number;
+	maxRetriesPerRequest?: number;
+	progressCallback?: (pctComplete: number) => void;
+}
+
+/**
+ *  A transaction uploader class that has been modified to handle uploading
+ *  the chunks of a transaction concurrently to the specified gateway url
+ *
+ * @example
+ *
+ *  ```ts
+ * 	await transaction.prepareChunks(transaction.data);
+ *
+ *		const transactionUploader = new MultiChunkTxUploader({
+ *			transaction,
+ *			gatewayUrl: new URL('https://arweave.net:443')
+ *		});
+ *
+ *		await transactionUploader.batchUploadChunks();
+ * ```
+ */
+export class MultiChunkTxUploader {
+	private chunkOffset = 0;
+	private txPosted = false;
+	private uploadedChunks = 0;
+	private hasFailedRequests = false;
+
+	public get isComplete(): boolean {
+		return this.txPosted && this.uploadedChunks === this.totalChunks;
+	}
+
+	public get totalChunks(): number {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		return this.transaction.chunks!.chunks.length;
+	}
+
+	public get pctComplete(): number {
+		return Math.trunc((this.uploadedChunks / this.totalChunks) * 100);
+	}
+
+	private gatewayUrl: URL;
+	private transaction: Transaction;
+	private maxConcurrentChunks: number;
+	private maxRetriesPerRequest: number;
+	private progressCallback?: (pctComplete: number) => void;
+
+	constructor({
+		gatewayUrl,
+		transaction,
+		maxConcurrentChunks = defaultMaxConcurrentChunks,
+		maxRetriesPerRequest = 8,
+		progressCallback
+	}: MultiChunkTxUploaderConstructorParams) {
+		if (!transaction.id) {
+			throw new Error(`Transaction is not signed`);
+		}
+		if (!transaction.chunks) {
+			throw new Error(`Transaction chunks not prepared`);
+		}
+
+		this.gatewayUrl = gatewayUrl;
+		this.transaction = transaction;
+		this.maxConcurrentChunks = maxConcurrentChunks;
+		this.maxRetriesPerRequest = maxRetriesPerRequest;
+		this.progressCallback = progressCallback;
+	}
+
+	/**
+	 * Uploads a transaction and all of its chunks until the upload is complete or has failed
+	 *
+	 * TODO: Convert this to a stream to report back event triggered progress
+	 *
+	 * @throws when any requests have failed beyond the maxRetries setting
+	 */
+	public async batchUploadChunks(): Promise<void> {
+		if (this.hasFailedRequests) {
+			throw new Error('Transaction upload has failed requests!');
+		}
+
+		if (!this.txPosted) {
+			await this.postTransactionHeader();
+
+			if (this.isComplete) {
+				return;
+			}
+		}
+
+		const numRemainingChunks = this.totalChunks - this.chunkOffset;
+		const numOfConcurrentUploadPromises = Math.min(numRemainingChunks, this.maxConcurrentChunks);
+
+		const uploadPromises: Promise<void>[] = [];
+		for (let index = 0; index < numOfConcurrentUploadPromises; index++) {
+			uploadPromises.push(this.uploadChunk());
+		}
+
+		try {
+			await Promise.all(uploadPromises);
+		} catch (err) {
+			throw new Error(err);
+		}
```

Comment:
Does it have different behavior than directly awaiting for `Promise.all`? The error would be thrown anyway...
Unless... the wrapping `Error` at the catch is doing something relevant, in that case:
```suggestion
		await Promise.all(uploadPromises).catch((err) => {
			throw new Error(err);
		})
```

Reply:
Removed this one: [4de6ca5](https://github.com/ardriveapp/ardrive-core-js/pull/179/commits/4de6ca56e6acdce8133df6d8b696e1e24b22ba0a)

Reply:
> However, I would advise against ever mixing async/await syntax with then/catch syntax.

Hmmm, are there any disadvantages by doing that? I find it completely safe and readable :opionion: